### PR TITLE
Add CLOSED-CAPTIONS= to writer for (STREAM-INF)

### PR DESCRIPTION
**Pull request by mbowBC**

### CODE REVIEW

**Comment by bradleyfalzon** [comment]
Would you be able to add reader support and tests?

File: reader_test.go [hunk 1]
```diff
@@ -96,7 +96,40 @@ func TestDecodeMasterPlaylistWithAlternatives(t *testing.T) {
...
+	// check parsed values
+	if p.ver != 4 {
+		t.Errorf("Version of parsed playlist = %d (must = 4)", p.ver)
+	}
+	if len(p.Variants) != 3 {
+		t.Fatal("not all variants in master playlist parsed")
+	}
+
+	for i, v := range p.Variants {
+		//spew.Dump(p.Variants)
```
Comment (bradleyfalzon) [inline reader_test.go]: Comment can be removed

File: reader_test.go [hunk 2]
```diff
@@ -96,7 +96,40 @@ func TestDecodeMasterPlaylistWithAlternatives(t *testing.T) {
...
+		if i == 0 && len(v.Alternatives) != 5 {
+			t.Fatalf("not all alternatives from #EXT-X-MEDIA parsed (has %d but should be 5", len(v.Alternatives))
+		}
+		if i > 0 && len(v.Alternatives) > 0 {
```
Comment (bradleyfalzon) [inline reader_test.go]: Neither of these `if` blocks belong here, looks like they were from the previous test testing Alternatives, feel free to just have the one `if v.Captions ~!= ...`, that's all you need to test for.

File: reader_test.go [hunk 3]
```diff
@@ -96,7 +96,40 @@ func TestDecodeMasterPlaylistWithAlternatives(t *testing.T) {
 			t.Fatal("should not be alternatives for this variant")
 		}
 	}
-	// fmt.Println(p.Encode().String())
+	//fmt.Println(p.Encode().String())
```
Comment (bradleyfalzon) [inline reader_test.go]: This line probably didn't need to be changed. We can remove it in another cleanup.

File: reader_test.go [hunk 4]
```diff
@@ -96,7 +96,40 @@ func TestDecodeMasterPlaylistWithAlternatives(t *testing.T) {
...
+			t.Fatalf("not all alternatives from #EXT-X-MEDIA parsed (has %d but should be 5", len(v.Alternatives))
+		}
+		if i > 0 && len(v.Alternatives) > 0 {
+			t.Fatal("should not be alternatives for this variant")
+		}
+		if v.Captions != "NONE" {
+			t.Fatal("variant field for CLOSED-CAPTIONS should be equal to NONE but it equals", v.Captions)
+		}
+	}
+	//fmt.Println(p.Encode().String())
```
Comment (bradleyfalzon) [inline reader_test.go]: This can be removed, it appears to have just come from a copy/paste.

File: writer_test.go [hunk 5]
```diff
@@ -550,6 +550,93 @@ func TestNewMasterPlaylistWithAlternatives(t *testing.T) {
 	if m.ver != 4 {
 		t.Fatalf("Expected version 4, actual, %d", m.ver)
 	}
+	fmt.Printf("%v\n", m)
+}
+
+// Create new master playlist supporting closed-caption=none
+func TestNewMasterPlaylistWithClosedCaptionEqNone(t *testing.T) {
```
Comment (bradleyfalzon) [inline writer_test.go]: This is a very comprehensive test, but I don't think it really needs to be. Can the same writing test be achieved by just having a single master playlist, with one variant setting the Alternative Closed Captions to none?

And then the test can be a straight forward strings.Contains, such as https://github.com/grafov/m3u8/blob/f55aa7a0ae5bfe63e82ee5adefa3aa4e8692b3dd/writer_test.go#L192-L221 ?

File: writer_test.go [hunk 6]
```diff
@@ -550,6 +550,93 @@ func TestNewMasterPlaylistWithAlternatives(t *testing.T) {
...
+	if err != nil {
+		t.Fatalf("Create media playlist failed: %s", err)
+	}
+	for i, vp := range vps {
+		m.Append(fmt.Sprintf("%d_rendition.m3u8", i+1), p, *vp)
+	}
+	if m.ver != 4 {
+		t.Fatalf("Expected version 4, actual, %d", m.ver)
+	}
+	// diff with master file using reader.
```
Comment (bradleyfalzon) [inline writer_test.go]: See comment on test, but I don't think this test needs to be any more complicated than a strings.Contains. 

If we do want this comprehensive test (@grafov thoughts?), then this test should be renamed to something like ` TestNewMasterPlaylistWithAlternates` and a simpler string comparison should be made instead of bringing in difflib.

Comment (grafov) [inline writer_test.go]: I think as the library code is simple and doesn't need external deps so the tests should avoid external deps too.

File: writer_test.go (hunk 5 above)
Comment (mbowBC) [inline writer_test.go]: ok will do.

**Comment by bradleyfalzon** [comment]
I can see more commits, and they're looking good, but you haven't stated you're finished, so I'll await your confirmation before reviewing.

**Comment by mbowBC** [comment]
I think that's it, let me know if I've missed anything.

File: writer_test.go [hunk 7]
```diff
@@ -550,6 +550,44 @@ func TestNewMasterPlaylistWithAlternatives(t *testing.T) {
 	if m.ver != 4 {
 		t.Fatalf("Expected version 4, actual, %d", m.ver)
 	}
+	fmt.Printf("%v\n", m)
+}
+
+// Create new master playlist supporting closed-caption=none
+func TestNewMasterPlaylistWithClosedCaptionEqNone(t *testing.T) {
+	m := NewMasterPlaylist()
+	var alts = []*Alternative{
```
Comment (bradleyfalzon) [inline writer_test.go]: One more thing, is this still required? If it's just to test for captions == none, I don't think we need an alternative for it? I'm not familiar with alternative or captions myself in m3u8 spec, so if an Alternative is required to set Captions to none then I'm OK with leaving it in, but if it's not required, then I think we can remove this from here too.

**Comment by bradleyfalzon** [comment]
This is fine, one minor comment, but then I think we can merge.

**Comment by mbowBC** [comment]
Thanks yes looks like don't need set up Alts, so have removed it.

**Comment by bradleyfalzon** [comment]
You're a champion, thanks for the time, looks good to me but the CI failure seems to be a gofmt issue, did you not have a chance to run that - could you?

**Comment by mbowBC** [comment]
apologies no idea why Gogland IDE didn't do that, cli way works fine.

**Comment by bradleyfalzon** [comment]
Thanks for the contribution @mbowBC
